Type parameter defines type member under inheritance with Scala 3.2.2 ?

## Observation
```Scala 
trait Trait :
  type Y
  var list: List[Y] = Nil

class Class[Y] extends Trait :
  def add(elm: Y): Unit = list = elm :: list

object Object extends Class[Int] :
  add(42)

println(Object.list)
```
compiles [under 3.2.2](https://scastie.scala-lang.org/0Ofm1YqwSRCuZJQRgn5i9Q) but not [under 3.2.1](https://scastie.scala-lang.org/d2uw1ZtcT0eWUT7FwS0KyQ). The latter gives the error: 
```
Found:    (elm : Y)
Required: Class.this.Y²

where:    Y  is a type in class Class
          Y² is a type in trait Trait
```
So it _looks like_ under 3.2.2 it is possible to define the type member from `Trait` with the type parameter from `Class`. That would be nice if this would fail to compile:
```Scala
trait Trait :
  type Y <: String
  var list: List[Y] = Nil

class Class[Y] extends Trait :
  def add(elm: Y): Unit = list = elm :: list

object Object extends Class[Int] :
  add(42)

println(Object.list)
```
where I changed `type Y` to `type Y <: String` but it does not. [This still produces `List(42)`](https://scastie.scala-lang.org/tEJzxGMxT8ionW37VUUiOw). 

## Expectation
Compiler error on the latter code OR the same behaviour as in Scala 3.2.1 of Scala 3.2.2 on the former code .

## Discussion
I observed this behaviour on the Jan, 13 and we discussed it om the [scala users forum under the title: 
Int <: String in Scala 3.2.2?](https://users.scala-lang.org/t/int-string-in-scala-3-2-2/9084). I had to wait though until Scastie with 3.2.2 was released for easy reference. I tested the latter code also under Scala [3.3.0-RC2](https://mvnrepository.com/artifact/org.scala-lang/scala3-library_3/3.3.0-RC2) and [3.3.1-RC1-bin-20230204-a356581-NIGHTLY](https://mvnrepository.com/artifact/org.scala-lang/scala3-library_3/3.3.1-RC1-bin-20230204-a356581-NIGHTLY) and they behave as 3.2.2 does.

Thank you for submitting. This appears to also be a pretty severe soundness issue, as I try to show in the reproduction below:
```scala
trait Trait:
  type Y <: String
  var list: List[Y] = Nil
  def listSubstr() =
    list.map(_.substring(0))

class Class[Y] extends Trait :
  def add(elm: Y): Unit = list = elm :: list

object Object extends Class[Int]:
  add(42)
  listSubstr()

@main def main() =
  println(Object.list)
```
compiles "successfully" on main (21729d296570726859a71a41fe0aba5a22d652c0) and 3.2.2, but when run will error out with a ClassCastException:

```scala
Exception in thread "main" java.lang.ExceptionInInitializerError
        at main$package$.main(main.scala:15)
        at main.main(main.scala:14)
Caused by: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String are in module java.base of loader 'bootstrap')
        at scala.collection.immutable.List.map(List.scala:246)
        at Trait.listSubstr(main.scala:5)
        at Trait.listSubstr$(main.scala:1)
        at Class.listSubstr(main.scala:7)
        at Object$.<clinit>(main.scala:12)
        ... 2 more
```

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -309,6 +309,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                 thirdTryNamed(tp2)
               else
                 (  (tp1.name eq tp2.name)
+                && !sym1.is(Private)
                 && tp2.isPrefixDependentMemberRef
                 && isSubPrefix(tp1.prefix, tp2.prefix)
                 && tp1.signature == tp2.signature

--- a/tests/neg/i16850.check
+++ b/tests/neg/i16850.check
@@ -1,0 +1,10 @@
+-- [E007] Type Mismatch Error: tests/neg/i16850.scala:7:33 -------------------------------------------------------------
+7 |  def add(elm: Y): Unit = list = elm :: list // error
+  |                                 ^^^
+  |                                 Found:    (elm : Y)
+  |                                 Required: Class.this.Y²
+  |
+  |                                 where:    Y  is a type in class Class
+  |                                           Y² is a type in trait Trait
+  |
+  | longer explanation available when compiling with `-explain`

--- a/tests/neg/i16850.scala
+++ b/tests/neg/i16850.scala
@@ -1,0 +1,10 @@
+
+trait Trait :
+  type Y
+  var list: List[Y] = Nil
+
+class Class[Y] extends Trait :
+  def add(elm: Y): Unit = list = elm :: list // error
+
+object Object extends Class[Int] :
+  add(42)

